File: backend/models/onepager.py
"""
OnePager Data Models
====================

Pydantic models for one-pager layouts and elements.
Based on validated POC schema from canva-poc/json_schema.py
"""

from pydantic import BaseModel, Field, validator
from typing import List, Optional, Dict, Any, Union, Literal
from enum import Enum
from datetime import datetime


class ElementType(str, Enum):
    """Types of elements in a one-pager."""
    HEADER = "header"
    HERO = "hero"
    FEATURES = "features"
    TESTIMONIALS = "testimonials"
    CTA = "cta"
    FOOTER = "footer"
    IMAGE = "image"
    TEXT_BLOCK = "text_block"
    # AI service-generated types
    HEADING = "heading"
    TEXT = "text"
    LIST = "list"
    BUTTON = "button"


class Dimensions(BaseModel):
    """Design dimensions in pixels."""
    width: int = Field(default=1080, ge=100, le=5000, description="Design width in pixels")
    height: int = Field(default=1920, ge=100, le=5000, description="Design height in pixels")


class Styling(BaseModel):
    """Element styling properties."""
    background_color: Optional[str] = Field(None, description="Hex color code (e.g., #0ea5e9)")
    text_color: Optional[str] = Field(None, description="Text color hex code")
    font_family: Optional[str] = Field(None, description="Font family name")
    font_size: Optional[int] = Field(None, ge=8, le=200, description="Font size in pixels")
    font_weight: Optional[str] = Field(None, description="Font weight (normal, bold, etc.)")
    text_align: Optional[str] = Field(None, description="Text alignment (left, center, right)")
    border_radius: Optional[int] = Field(None, description="Border radius in pixels")
    padding: Optional[int] = Field(None, description="Internal padding in pixels")
    
    @validator('background_color', 'text_color')
    def validate_color(cls, v):
        """Validate color format is hex."""
        if v and not (v.startswith('#') and len(v) == 7):
            raise ValueError("Color must be hex format: #RRGGBB")
        return v


class Position(BaseModel):
    """Element position in pixels."""
    x: int = Field(description="X coordinate in pixels")
    y: int = Field(description="Y coordinate in pixels")
    width: int = Field(description="Element width in pixels")
    height: int = Field(description="Element height in pixels")


# Element content models (flexible Dict for now, will expand later)
class HeaderContent(BaseModel):
    """Content for HEADER elements."""
    title: str
    subtitle: Optional[str] = None
    logo_url: Optional[str] = None


class HeroContent(BaseModel):
    """Content for HERO elements."""
    headline: str
    subheadline: Optional[str] = None
    description: str
    image_url: Optional[str] = None
    cta_text: Optional[str] = None
    cta_url: Optional[str] = None


class CTAContent(BaseModel):
    """Content for CTA elements."""
    primary_text: str
    primary_url: str
    secondary_text: Optional[str] = None
    secondary_url: Optional[str] = None
    supporting_text: Optional[str] = None


class FooterContent(BaseModel):
    """Content for FOOTER elements."""
    company_name: str
    contact_email: Optional[str] = None
    contact_phone: Optional[str] = None
    website_url: Optional[str] = None
    copyright_text: Optional[str] = None


class ImageContent(BaseModel):
    """Content for IMAGE elements."""
    image_url: str
    alt_text: Optional[str] = None
    caption: Optional[str] = None


class TextBlockContent(BaseModel):
    """Content for TEXT_BLOCK elements."""
    text: str
    heading_level: Optional[int] = Field(None, ge=1, le=6, description="Heading level (1-6)")


class OnePagerElement(BaseModel):
    """Single element in one-pager layout."""
    id: str = Field(description="Unique element identifier")
    type: ElementType = Field(description="Element type")
    title: Optional[str] = Field(None, description="Element title (for sections)")
    content: Union[str, List[str], Dict[str, Any]] = Field(
        description="Element content (text string, list of items, or structured dict)"
    )
    styling: Optional[Styling] = Field(None, description="Element styling overrides")
    position: Optional[Position] = Field(None, description="Element position (if absolute)")
    order: int = Field(description="Display order within the layout")
    visible: bool = Field(default=True, description="Whether element is visible")

    class Config:
        use_enum_values = True


class OnePagerLayout(BaseModel):
    """Complete one-pager layout specification."""
    title: str = Field(description="One-pager title/name")
    description: Optional[str] = Field(None, description="One-pager description")
    dimensions: Dimensions = Field(default_factory=Dimensions, description="Design dimensions")
    elements: List[OnePagerElement] = Field(description="List of design elements")
    version: int = Field(default=1, description="Layout version number")
    created_at: Optional[datetime] = Field(None, description="Creation timestamp")
    updated_at: Optional[datetime] = Field(None, description="Last update timestamp")
    
    # Brand profile reference (will connect to Profile model later in Phase 2)
    brand_colors: Optional[Dict[str, str]] = Field(None, description="Brand color overrides")
    brand_fonts: Optional[Dict[str, str]] = Field(None, description="Brand font overrides")
    brand_logo_url: Optional[str] = Field(None, description="Brand logo URL")
    
    @validator('elements')
    def validate_element_order(cls, v):
        """Ensure element order values are unique and sequential."""
        orders = [element.order for element in v]
        if len(orders) != len(set(orders)):
            raise ValueError("Element order values must be unique")
        return v
    
    def get_element_by_id(self, element_id: str) -> Optional[OnePagerElement]:
        """Get an element by its ID."""
        for element in self.elements:
            if element.id == element_id:
                return element
        return None
    
    def get_elements_by_type(self, element_type: ElementType) -> List[OnePagerElement]:
        """Get all elements of a specific type."""
        return [element for element in self.elements if element.type == element_type]


# Helper Functions for API Responses

def onepager_helper(onepager_doc: Dict[str, Any]) -> Dict[str, Any]:
    """
    Transform MongoDB onepager document to API response format.

    Args:
        onepager_doc: Raw MongoDB document

    Returns:
        Dictionary ready for OnePagerResponse model
    """
    return {
        "id": str(onepager_doc["_id"]),
        "user_id": str(onepager_doc["user_id"]),
        "brand_kit_id": str(onepager_doc["brand_kit_id"]) if onepager_doc.get("brand_kit_id") else None,
        "title": onepager_doc["title"],
        "status": onepager_doc["status"],
        "content": onepager_doc.get("content", {}),
        "layout": onepager_doc.get("layout", []),
        "style_overrides": onepager_doc.get("style_overrides", {}),
        "generation_metadata": onepager_doc.get("generation_metadata", {}),
        "version_history": onepager_doc.get("version_history", []),
        "layout_params": onepager_doc.get("layout_params"),
        "design_rationale": onepager_doc.get("design_rationale"),
        "created_at": onepager_doc["created_at"],
        "updated_at": onepager_doc["updated_at"],
        "last_accessed": onepager_doc.get("last_accessed", onepager_doc["updated_at"])
    }


def onepager_summary_helper(onepager_doc: Dict[str, Any]) -> Dict[str, Any]:
    """
    Transform MongoDB onepager document to summary format for list views.

    Args:
        onepager_doc: Raw MongoDB document

    Returns:
        Dictionary ready for OnePagerSummary model
    """
    return {
        "id": str(onepager_doc["_id"]),
        "title": onepager_doc["title"],
        "status": onepager_doc["status"],
        "created_at": onepager_doc["created_at"],
        "updated_at": onepager_doc["updated_at"],
        "has_brand_kit": bool(onepager_doc.get("brand_kit_id"))
    }


# ==========================================
# Layout Parameters Models
# ==========================================
# New data models for AI-driven layout customization
# Enables iterative design refinement beyond content


class ColorScheme(BaseModel):
    """
    Color scheme configuration for one-pager design.

    All colors must be in hex format (#RRGGBB).
    Used for dynamic PDF template styling.
    """
    primary: str = Field(
        default="#1568B8",
        description="Primary brand color (hex format)"
    )
    secondary: str = Field(
        default="#864CBD",
        description="Secondary brand color (hex format)"
    )
    accent: str = Field(
        default="#FF6B6B",
        description="Accent color for highlights (hex format)"
    )
    text_primary: str = Field(
        default="#1A202C",
        description="Primary text color (hex format)"
    )
    background: str = Field(
        default="#FFFFFF",
        description="Background color (hex format)"
    )

    @validator('primary', 'secondary', 'accent', 'text_primary', 'background')
    def validate_hex_color(cls, v):
        """Validate color format is hex (#RRGGBB)."""
        if not (v.startswith('#') and len(v) == 7):
            raise ValueError("Color must be in hex format: #RRGGBB")
        try:
            int(v[1:], 16)  # Validate hex characters
        except ValueError:
            raise ValueError("Invalid hex color code")
        return v


class Typography(BaseModel):
    """
    Typography parameters for scaling and font configuration.

    Scale values allow AI to adjust text sizes based on content characteristics.
    """
    heading_font: str = Field(
        default="Inter",
        description="Font family for headings"
    )
    body_font: str = Field(
        default="Inter",
        description="Font family for body text"
    )
    h1_scale: float = Field(
        default=1.0,
        ge=0.8,
        le=1.5,
        description="H1 size multiplier (0.8-1.5)"
    )
    h2_scale: float = Field(
        default=1.0,
        ge=0.8,
        le=1.5,
        description="H2 size multiplier (0.8-1.5)"
    )
    body_scale: float = Field(
        default=1.0,
        ge=0.8,
        le=1.3,
        description="Body text size multiplier (0.8-1.3)"
    )
    line_height_scale: float = Field(
        default=1.0,
        ge=0.8,
        le=1.4,
        description="Line height multiplier (0.8-1.4)"
    )


class Spacing(BaseModel):
    """
    Spacing configuration for section gaps and padding.

    Controls vertical spacing between sections and overall padding density.
    """
    section_gap: Literal["tight", "normal", "loose"] = Field(
        default="normal",
        description="Vertical spacing between sections"
    )
    padding_scale: float = Field(
        default=1.0,
        ge=0.5,
        le=2.0,
        description="Overall padding multiplier (0.5-2.0)"
    )


class SectionLayout(BaseModel):
    """
    Layout configuration for an individual section.

    Defines grid structure and alignment for section content.
    """
    columns: Literal[1, 2, 3] = Field(
        default=1,
        description="Number of columns (1-3)"
    )
    alignment: Literal["left", "center", "right"] = Field(
        default="left",
        description="Content alignment"
    )
    image_position: Optional[Literal["top", "left", "right", "none"]] = Field(
        default="top",
        description="Image placement relative to content"
    )


class LayoutParams(BaseModel):
    """
    Complete layout parameter configuration for one-pager.

    Enables AI-driven design customization beyond content.
    All parameters have sensible defaults and validation ranges.
    """
    color_scheme: ColorScheme = Field(
        default_factory=ColorScheme,
        description="Color palette configuration"
    )
    typography: Typography = Field(
        default_factory=Typography,
        description="Typography scaling configuration"
    )
    spacing: Spacing = Field(
        default_factory=Spacing,
        description="Spacing and padding configuration"
    )
    section_layouts: Dict[str, SectionLayout] = Field(
        default_factory=dict,
        description="Per-section layout configuration (key: section name)"
    )

    class Config:
        json_schema_extra = {
            "example": {
                "color_scheme": {
                    "primary": "#1568B8",
                    "secondary": "#864CBD",
                    "accent": "#FF6B6B",
                    "text_primary": "#1A202C",
                    "background": "#FFFFFF"
                },
                "typography": {
                    "heading_font": "Inter",
                    "body_font": "Inter",
                    "h1_scale": 1.2,
                    "h2_scale": 1.0,
                    "body_scale": 1.0,
                    "line_height_scale": 1.1
                },
                "spacing": {
                    "section_gap": "normal",
                    "padding_scale": 1.0
                },
                "section_layouts": {
                    "features": {
                        "columns": 2,
                        "alignment": "left"
                    },
                    "benefits": {
                        "columns": 1,
                        "alignment": "center"
                    }
                }
            }
        }


def get_default_layout_params() -> LayoutParams:
    """
    Get default layout parameters with sensible styling.

    Returns:
        LayoutParams instance with all default values

    Usage:
        Used for onepagers without custom layout configuration.
        Provides baseline styling that works for most content.
    """
    return LayoutParams(
        color_scheme=ColorScheme(),
        typography=Typography(),
        spacing=Spacing(),
        section_layouts={
            "features": SectionLayout(columns=2, alignment="left"),
            "benefits": SectionLayout(columns=1, alignment="center"),
            "integrations": SectionLayout(columns=3, alignment="left")
        }
    )


def validate_layout_params(data: Dict[str, Any]) -> Optional[LayoutParams]:
    """
    Validate layout parameters data and return LayoutParams instance if valid.

    Args:
        data: Dictionary containing layout parameter data

    Returns:
        LayoutParams instance if validation succeeds, None if validation fails

    Usage:
        Used by API endpoints to validate user-submitted layout parameters.
        Returns None on validation errors instead of raising exceptions.

    Example:
        >>> params_dict = {"color_scheme": {"primary": "#FF0000"}}
        >>> validated = validate_layout_params(params_dict)
        >>> if validated:
        ...     # Use validated params
        ...     pass
    """
    try:
        return LayoutParams(**data)
    except Exception:
        return None


def merge_layout_params(
    base: Optional[LayoutParams] = None,
    updates: Optional[Dict[str, Any]] = None
) -> LayoutParams:
    """
    Merge user layout parameter updates with base parameters.

    Args:
        base: Base LayoutParams (defaults to get_default_layout_params() if None)
        updates: Dictionary of user updates to apply

    Returns:
        New LayoutParams instance with merged values

    Usage:
        Used when user modifies specific layout parameters while keeping others default.
        Enables incremental customization without requiring full parameter specification.

    Example:
        >>> base = get_default_layout_params()
        >>> updates = {"spacing": {"section_gap": "loose"}}
        >>> merged = merge_layout_params(base, updates)
        >>> # Result: All defaults except section_gap is now "loose"
    """
    # Get base params (use defaults if not provided)
    if base is None:
        base = get_default_layout_params()

    # If no updates, return base as-is
    if not updates:
        return base

    # Convert base to dict for merging
    base_dict = base.dict()

    # Deep merge updates into base
    for key, value in updates.items():
        if key in base_dict:
            # If both are dicts, merge recursively
            if isinstance(value, dict) and isinstance(base_dict[key], dict):
                base_dict[key] = {**base_dict[key], **value}
            else:
                # Otherwise, replace value
                base_dict[key] = value
        else:
            # New key, add directly
            base_dict[key] = value

    # Validate and return merged params
    try:
        return LayoutParams(**base_dict)
    except Exception:
        # If merged params are invalid, return base unchanged
        return base

File: frontend/src/types/onepager.ts
/**
 * OnePager Type Definitions
 *
 * Type-safe interfaces for OnePager data structures.
 * Mirrors backend Pydantic schemas from backend/onepagers/schemas.py
 */

/**
 * Polymorphic content type for sections
 * Handles string (heading, text, button), array (list), or object (hero, features)
 */
export type SectionContentType =
  | string
  | string[]
  | {
      headline?: string;
      subheadline?: string;
      description?: string;
      items?: Array<{ title: string; description: string }>;
      [key: string]: any;
    };

/**
 * Section types supported by the renderer
 * Matches backend ElementType enum
 */
export type SectionType =
  | 'heading'
  | 'text'
  | 'list'
  | 'button'
  | 'hero'
  | 'features'
  | 'testimonials'
  | 'cta'
  | 'footer'
  | 'image';

/**
 * OnePager status values
 */
export type OnePagerStatus = 'wireframe' | 'draft' | 'published' | 'archived';

/**
 * PDF template options for export
 * Maps to backend PDFTemplate enum
 */
export type PDFTemplate = 'minimalist' | 'bold' | 'business' | 'product';

/**
 * Content section within a OnePager
 * Represents a single block of content (e.g., heading, text, list)
 */
export interface ContentSection {
  id: string;
  type: SectionType;
  title?: string;
  content: SectionContentType;
  order: number;
}

/**
 * OnePager content structure
 * Top-level content container with headline, sections, and structured fields
 */
export interface OnePagerContent {
  headline: string;
  subheadline?: string;
  sections: ContentSection[];

  // Structured content fields
  problem?: string;
  solution?: string;
  features?: string[];
  benefits?: string[];
  integrations?: string[];
  social_proof?: string;
  cta?: {
    text: string;
    url: string;
  };
  visuals?: Array<{
    url: string;
    type?: string;
    alt_text?: string;
  }>;
}

/**
 * Layout block positioning information
 * Used for absolute positioning in canvas (future enhancement)
 */
export interface LayoutBlock {
  block_id: string;
  type: string;
  position: {
    x: number;
    y: number;
  };
  size: {
    width: string | number;
    height: string | number;
  };
  order: number;
}

/**
 * AI generation metadata
 * Tracks prompts, iterations, and AI model used
 */
export interface GenerationMetadata {
  prompts: string[];
  iterations: number;
  ai_model: string;
  last_generated_at: string;
}

/**
 * Version history snapshot
 * Captures state at a specific point in time
 * Includes layout_params to restore complete design state
 */
export interface VersionSnapshot {
  version: number;
  content: OnePagerContent;
  layout: LayoutBlock[];
  layout_params?: LayoutParams | null;  // Added for complete state restoration
  created_at: string;
  change_description?: string;  // Changed from 'description' to match backend
}

/**
 * LayoutParams Type Definitions
 * ===============================
 * Types for AI-driven layout customization beyond content.
 * Mirrors backend LayoutParams model from backend/models/onepager.py
 */

/**
 * Color scheme configuration
 * All colors must be in hex format (#RRGGBB)
 */
export interface ColorScheme {
  primary: string;
  secondary: string;
  accent: string;
  text_primary: string;
  background: string;
}

/**
 * Typography scaling configuration
 * All scales have validation ranges (see backend model)
 */
export interface Typography {
  heading_font: string;
  body_font: string;
  h1_scale: number;          // 0.8 - 1.5
  h2_scale: number;          // 0.8 - 1.5
  body_scale: number;        // 0.8 - 1.3
  line_height_scale: number; // 0.8 - 1.4
}

/**
 * Spacing configuration
 * Controls section gaps and padding
 */
export interface Spacing {
  section_gap: 'tight' | 'normal' | 'loose';
  padding_scale: number; // 0.5 - 2.0
}

/**
 * Section-specific layout configuration
 * Defines columns, alignment, and image positioning
 */
export interface SectionLayout {
  columns: 1 | 2 | 3;
  alignment: 'left' | 'center' | 'right';
  image_position?: 'top' | 'left' | 'right' | 'none' | null;
}

/**
 * Complete layout parameters
 * Container for all design customization settings
 */
export interface LayoutParams {
  color_scheme: ColorScheme;
  typography: Typography;
  spacing: Spacing;
  section_layouts: Record<string, SectionLayout>;
}

/**
 * Complete OnePager data structure
 * Full representation with all metadata
 */
export interface OnePager {
  id: string;
  user_id: string;
  brand_kit_id?: string;
  title: string;
  status: OnePagerStatus;
  content: OnePagerContent;
  layout: LayoutBlock[];
  style_overrides: Record<string, any>;
  generation_metadata: GenerationMetadata;
  version_history: VersionSnapshot[];
  layout_params?: LayoutParams | null;
  design_rationale?: string | null;
  pdf_template: PDFTemplate;  // PDF template for export
  created_at: string;
  updated_at: string;
  last_accessed: string;
}

/**
 * OnePager summary for list views
 * Lightweight version without full content
 */
export interface OnePagerSummary {
  id: string;
  title: string;
  status: OnePagerStatus;
  created_at: string;
  updated_at: string;
  has_brand_kit: boolean;
}

/**
 * Call-to-action data
 */
export interface CTAData {
  text: string;
  url: string;
}

/**
 * Visual/image data
 */
export interface VisualData {
  url: string;
  type?: string;
  alt_text?: string;
}

/**
 * Request payload for creating a new OnePager
 * Enhanced with structured fields from Week 2 MVP requirements
 */
export interface OnePagerCreateData {
  title: string;

  // Product selection
  product_id?: string;

  // Core content fields (required)
  problem: string;
  solution: string;
  features: string[];
  benefits: string[];

  // Optional fields
  integrations?: string[];
  social_proof?: string;
  cta: CTAData;
  visuals?: VisualData[];

  // Brand and audience
  brand_kit_id?: string;
  target_audience?: string;

  // AI generation (optional)
  input_prompt?: string;
}

/**
 * Request payload for AI iterative refinement
 */
export interface OnePagerIterateData {
  feedback: string;
  iteration_type?: 'content' | 'layout' | 'both';
  preserve_elements?: string[];
}

/**
 * Request payload for updating OnePager metadata
 */
export interface OnePagerUpdateData {
  title?: string;
  status?: OnePagerStatus;
  style_overrides?: Record<string, any>;
  brand_kit_id?: string;
  pdf_template?: PDFTemplate;
}

/**
 * PDF export format options
 */
export type PDFFormat = 'letter' | 'a4' | 'tabloid';

/**
 * Response from suggest-layout API endpoint
 * AI suggestions for layout parameters without modifying content
 */
export interface LayoutSuggestionResponse {
  suggested_layout_params: LayoutParams;
  design_rationale: string;
}

/**
 * Helper function to get default layout parameters
 * Mirrors backend get_default_layout_params()
 */
export function getDefaultLayoutParams(): LayoutParams {
  return {
    color_scheme: {
      primary: '#1568B8',
      secondary: '#864CBD',
      accent: '#FF6B6B',
      text_primary: '#1A202C',
      background: '#FFFFFF',
    },
    typography: {
      heading_font: 'Inter',
      body_font: 'Inter',
      h1_scale: 1.0,
      h2_scale: 1.0,
      body_scale: 1.0,
      line_height_scale: 1.0,
    },
    spacing: {
      section_gap: 'normal',
      padding_scale: 1.0,
    },
    section_layouts: {
      features: {
        columns: 2,
        alignment: 'left',
        image_position: 'top',
      },
      benefits: {
        columns: 1,
        alignment: 'center',
        image_position: 'top',
      },
      integrations: {
        columns: 3,
        alignment: 'left',
        image_position: 'top',
      },
    },
  };
}
